File: typescript/src/oauth/exchange.ts
/**
 * OAuth 2.0 token exchange and refresh for Basecamp SDK.
 *
 * Handles authorization code exchange and token refresh operations.
 * Supports both standard OAuth 2.0 and Basecamp's Launchpad legacy format.
 */

import { MAX_TOKEN_LIFETIME_SECONDS, raceAbort, resolveRequestTimeoutMs } from "./limits.js";
import { BasecampError } from "../errors.js";
import { isLocalhost } from "../security.js";
import type {
  ExchangeRequest,
  RefreshRequest,
  OAuthToken,
  RawTokenResponse,
  OAuthErrorResponse,
} from "./types.js";

/** Default per-request timeout (ms) for a token exchange/refresh round trip. */
const DEFAULT_TOKEN_TIMEOUT_MS = 30_000;

/**
 * The redirects the token endpoint is refused (SPEC §16 "Token-Endpoint
 * Transport Policy") — the same set the signed download hop refuses (§14).
 * 304 is not in the set: it is a cache validator, not a redirect-with-
 * Location, and falls through to the generic non-ok handling below.
 */
const REDIRECT_STATUSES = [301, 302, 303, 307, 308];

/**
 * Options for token exchange/refresh operations.
 */
export interface TokenOptions {
  /** Custom fetch function for testing or custom HTTP handling */
  fetch?: typeof globalThis.fetch;
  /** Request timeout in milliseconds (default: 30000) */
  timeoutMs?: number;
}

/**
 * Exchanges an authorization code for access and refresh tokens.
 *
 * Supports both standard OAuth 2.0 and Basecamp's Launchpad legacy format.
 * Use `useLegacyFormat: true` for Launchpad compatibility.
 *
 * @param request - Exchange request parameters
 * @param options - Optional configuration
 * @returns The token response
 * @throws BasecampError on validation, network, or authentication errors
 *
 * @example
 * ```ts
 * // Standard OAuth 2.0
 * const token = await exchangeCode({
 *   tokenEndpoint: config.tokenEndpoint,
 *   code: "auth_code_from_callback",
 *   redirectUri: "https://myapp.com/callback",
 *   clientId: "my_client_id",
 *   clientSecret: "my_client_secret",
 * });
 *
 * // Launchpad legacy format
 * const token = await exchangeCode({
 *   tokenEndpoint: "https://launchpad.37signals.com/authorization/token",
 *   code: "auth_code",
 *   redirectUri: "https://myapp.com/callback",
 *   clientId: "my_client_id",
 *   clientSecret: "my_client_secret",
 *   useLegacyFormat: true,
 * });
 * ```
 */
export async function exchangeCode(
  request: ExchangeRequest,
  options: TokenOptions = {}
): Promise<OAuthToken> {
  // Validate required fields
  if (!request.tokenEndpoint) {
    throw new BasecampError("validation", "Token endpoint is required");
  }
  if (!request.code) {
    throw new BasecampError("validation", "Authorization code is required");
  }
  if (!request.redirectUri) {
    throw new BasecampError("validation", "Redirect URI is required");
  }
  if (!request.clientId) {
    throw new BasecampError("validation", "Client ID is required");
  }

  // Build request body
  const body = new URLSearchParams();

  if (request.useLegacyFormat) {
    // Launchpad uses non-standard "type" parameter
    body.set("type", "web_server");
  } else {
    // Standard OAuth 2.0
    body.set("grant_type", "authorization_code");
  }

  body.set("code", request.code);
  body.set("redirect_uri", request.redirectUri);
  body.set("client_id", request.clientId);

  if (request.clientSecret) {
    body.set("client_secret", request.clientSecret);
  }
  if (request.codeVerifier) {
    body.set("code_verifier", request.codeVerifier);
  }

  return doTokenRequest(request.tokenEndpoint, body, options);
}

/**
 * Refreshes an access token using a refresh token.
 *
 * Supports both standard OAuth 2.0 and Basecamp's Launchpad legacy format.
 * Use `useLegacyFormat: true` for Launchpad compatibility.
 *
 * @param request - Refresh request parameters
 * @param options - Optional configuration
 * @returns The new token response
 * @throws BasecampError on validation, network, or authentication errors
 *
 * @example
 * ```ts
 * // Standard OAuth 2.0
 * const newToken = await refreshToken({
 *   tokenEndpoint: config.tokenEndpoint,
 *   refreshToken: oldToken.refreshToken,
 *   clientId: "my_client_id",
 *   clientSecret: "my_client_secret",
 * });
 *
 * // Launchpad legacy format
 * const newToken = await refreshToken({
 *   tokenEndpoint: "https://launchpad.37signals.com/authorization/token",
 *   refreshToken: oldToken.refreshToken,
 *   useLegacyFormat: true,
 * });
 * ```
 */
export async function refreshToken(
  request: RefreshRequest,
  options: TokenOptions = {}
): Promise<OAuthToken> {
  // Validate required fields
  if (!request.tokenEndpoint) {
    throw new BasecampError("validation", "Token endpoint is required");
  }
  if (!request.refreshToken) {
    throw new BasecampError("validation", "Refresh token is required");
  }

  // Build request body
  const body = new URLSearchParams();

  if (request.useLegacyFormat) {
    // Launchpad uses non-standard "type" parameter
    body.set("type", "refresh");
  } else {
    // Standard OAuth 2.0
    body.set("grant_type", "refresh_token");
  }

  body.set("refresh_token", request.refreshToken);

  if (request.clientId) {
    body.set("client_id", request.clientId);
  }
  if (request.clientSecret) {
    body.set("client_secret", request.clientSecret);
  }
  if (request.resource) {
    body.set("resource", request.resource);
  }

  return doTokenRequest(request.tokenEndpoint, body, options);
}

/**
 * Validates that a URL uses the HTTPS scheme (allows localhost for testing).
 */
function requireHTTPS(url: string, label: string): void {
  try {
    const parsed = new URL(url);
    if (parsed.protocol !== "https:" && !isLocalhost(parsed.hostname)) {
      throw new BasecampError("validation", `${label} must use HTTPS: ${url}`);
    }
  } catch (err) {
    if (err instanceof BasecampError) throw err;
    throw new BasecampError("validation", `Invalid ${label}: ${url}`);
  }
}

/**
 * Parses Content-Length header, returning null for missing/invalid values.
 * Treats non-numeric or negative values as invalid (fail closed).
 */
function parseContentLength(header: string | null): number | null {
  if (!header) return null;
  const parsed = parseInt(header, 10);
  // NaN, negative, or non-integer values are invalid
  if (!Number.isInteger(parsed) || parsed < 0) return null;
  // Reject if the header contains non-digit characters (e.g., "123abc")
  if (!/^\d+$/.test(header.trim())) return null;
  return parsed;
}

/**
 * Reads a response body with streaming byte limit.
 * Uses true byte counting even when Content-Length is absent or inaccurate.
 */
async function readResponseWithByteLimit(
  response: Response,
  maxBytes: number,
  httpStatus: number
): Promise<string> {
  // Parse Content-Length, treating invalid values as missing (fail closed)
  const contentLengthBytes = parseContentLength(
    response.headers.get("Content-Length")
  );

  // Fast path: reject if Content-Length exceeds limit
  if (contentLengthBytes !== null && contentLengthBytes > maxBytes) {
    throw new BasecampError(
      "api_error",
      `Token response too large (Content-Length: ${contentLengthBytes} bytes, max ${maxBytes})`,
      { httpStatus }
    );
  }

  // If no body or body isn't streamable, require valid Content-Length
  if (!response.body) {
    // If Content-Length is missing or invalid, we cannot protect against DoS - fail closed.
    // If it was valid and within limits, we can safely read.
    if (contentLengthBytes === null) {
      throw new BasecampError(
        "api_error",
        "Cannot safely read token response: no valid Content-Length header and streaming unavailable",
        { httpStatus }
      );
    }
    // Content-Length was valid and within limits, safe to read
    return response.text();
  }

  // Stream with byte counting
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let totalBytes = 0;

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      totalBytes += value.length;
      if (totalBytes > maxBytes) {
        reader.cancel();
        throw new BasecampError(
          "api_error",
          `Token response too large (${totalBytes}+ bytes, max ${maxBytes})`,
          { httpStatus }
        );
      }
      chunks.push(value);
    }
  } finally {
    reader.releaseLock();
  }

  // Concatenate and decode
  const combined = new Uint8Array(totalBytes);
  let offset = 0;
  for (const chunk of chunks) {
    combined.set(chunk, offset);
    offset += chunk.length;
  }

  return new TextDecoder().decode(combined);
}

/**
 * Performs the actual HTTP token request.
 */
async function doTokenRequest(
  tokenEndpoint: string,
  body: URLSearchParams,
  options: TokenOptions
): Promise<OAuthToken> {
  requireHTTPS(tokenEndpoint, "token endpoint");

  const { fetch: customFetch = globalThis.fetch, timeoutMs = DEFAULT_TOKEN_TIMEOUT_MS } = options;

  // Create abort controller for timeout. The clamp closes the unbounded
  // hole an unvalidated timeoutMs left open: NaN/Infinity made setTimeout
  // fire at ~1 ms (an instant abort masquerading as a network failure), and
  // an oversized finite value could hold a stalled request open for weeks.
  const controller = new AbortController();
  const timeoutId = setTimeout(
    () => controller.abort(),
    resolveRequestTimeoutMs(timeoutMs, DEFAULT_TOKEN_TIMEOUT_MS)
  );

  try {
    // The whole round trip runs INSIDE a race against the controller's
    // signal, like the device flow's POSTs: a cooperative fetch rejects on
    // abort anyway, but a custom fetch that ignores its AbortSignal must not
    // hold the exchange past its timeout — the race rejects the moment the
    // timeout fires, and a late settlement is discarded.
    const { response, responseText } = await raceAbort(controller.signal, async () => {
      const raced = await customFetch(tokenEndpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          Accept: "application/json",
        },
        body: body.toString(),
        signal: controller.signal,
        // Never chase an attacker-influenced Location: the token endpoint may
        // come from discovered metadata, and a followed 307/308 would re-POST
        // the credentials wherever it points (SPEC §16).
        redirect: "manual",
      });

      // A browser fetch answers redirect: "manual" with an opaqueredirect —
      // type "opaqueredirect", status 0, no headers — rather than the 3xx
      // Node hands back, so the status table below cannot see it. Refuse it
      // by type; the real status is hidden, so the error carries none.
      if (raced.type === "opaqueredirect") {
        void raced.body?.cancel().catch(() => {});
        throw new BasecampError(
          "api_error",
          "redirect on the token endpoint is not followed (status hidden by the browser's opaque redirect)"
        );
      }

      // A suppressed redirect is refused by status BEFORE any body read, so a
      // 3xx whose body stalls forever cannot degrade into a timeout. Release
      // the unread stream (non-blocking) so refusals don't retain sockets.
      if (REDIRECT_STATUSES.includes(raced.status)) {
        void raced.body?.cancel().catch(() => {});
        throw new BasecampError(
          "api_error",
          `redirect ${raced.status} on the token endpoint is not followed`,
          { httpStatus: raced.status }
        );
      }

      const MAX_TOKEN_RESPONSE_BYTES = 1 * 1024 * 1024; // 1 MB

      // Use streaming reader with true byte-level limit enforcement.
      // This handles cases where Content-Length is absent or inaccurate.
      const racedText = await readResponseWithByteLimit(
        raced,
        MAX_TOKEN_RESPONSE_BYTES,
        raced.status
      );
      return { response: raced, responseText: racedText };
    });

    let data: RawTokenResponse | OAuthErrorResponse;

    try {
      data = JSON.parse(responseText);
    } catch {
      // A token response that fails to parse may still contain credential
      // material (a syntactically-broken body carrying an access_token) —
      // never echo ANY of it into an error message, where it would reach
      // logs and exception telemetry. The status is diagnosis enough.
      throw new BasecampError("api_error", "Failed to parse token response", {
        httpStatus: response.status,
      });
    }

    // A valid-JSON-but-non-object body (null, array, number, string) is a
    // malformed response on EVERY status — the error branch below would
    // otherwise deref null and surface a raw TypeError misclassified as
    // retryable network. Fail as api_error carrying the HTTP status.
    if (typeof data !== "object" || data === null || Array.isArray(data)) {
      throw new BasecampError("api_error", "Token response is not a JSON object", {
        httpStatus: response.status,
      });
    }

    // Check for error response
    if (!response.ok) {
      const errorData = data as OAuthErrorResponse;
      // Non-string error/error_description (numbers, objects) must not throw
      // a raw TypeError below — that would be misclassified as retryable
      // network, losing the api_error status context.
      const rawMessage =
        (typeof errorData.error_description === "string" && errorData.error_description) ||
        (typeof errorData.error === "string" && errorData.error) ||
        "Token request failed";
      const message = rawMessage.length > 500 ? rawMessage.slice(0, 497) + "..." : rawMessage;

      if (response.status === 401 || errorData.error === "invalid_grant") {
        throw new BasecampError("auth_required", message, {
          httpStatus: response.status,
          hint: "The authorization code or refresh token may be invalid or expired",
        });
      }

      throw new BasecampError("api_error", message, {
        httpStatus: response.status,
      });
    }

    // Parse successful response
    const tokenData = data as RawTokenResponse;

    // Non-empty STRING, not merely truthy: a numeric access_token is not a
    // usable credential. Carry the HTTP status like every other malformed-
    // response raise so failures are diagnosable.
    if (typeof tokenData.access_token !== "string" || tokenData.access_token === "") {
      throw new BasecampError("api_error", "Token response missing or non-string access_token", {
        httpStatus: response.status,
      });
    }

    // Absent/null token_type defaults to Bearer, but a present-but-empty (or
    // non-string) one is a malformed response — matching the stricter
    // device-flow validation rather than silently coercing "" to Bearer.
    if (tokenData.token_type != null && (typeof tokenData.token_type !== "string" || tokenData.token_type === "")) {
      throw new BasecampError("api_error", "Token response token_type must be a non-empty string when present", {
        httpStatus: response.status,
      });
    }

    // The remaining optional fields get the device-flow strictness: a
    // non-string refresh_token/scope or a non-finite/fractional/oversized
    // expires_in would leak malformed values through the public OAuthToken
    // type (or build an Invalid Date expiry).
    if (tokenData.refresh_token != null && typeof tokenData.refresh_token !== "string") {
      throw new BasecampError("api_error", "Token response refresh_token must be a string", {
        httpStatus: response.status,
      });
    }
    if (tokenData.scope != null && typeof tokenData.scope !== "string") {
      throw new BasecampError("api_error", "Token response scope must be a string", {
        httpStatus: response.status,
      });
    }
    if (
      tokenData.expires_in != null &&
      (typeof tokenData.expires_in !== "number" ||
        !Number.isInteger(tokenData.expires_in) ||
        tokenData.expires_in <= 0 ||
        tokenData.expires_in > MAX_TOKEN_LIFETIME_SECONDS)
    ) {
      throw new BasecampError(
        "api_error",
        `Token response expires_in must be a finite positive whole number no greater than ${MAX_TOKEN_LIFETIME_SECONDS} seconds`,
        { httpStatus: response.status }
      );
    }

    // resource: absent and JSON null are unset; when present it must be a
    // non-empty string (SPEC §16) — an empty binding is not a binding.
    if (tokenData.resource != null && (typeof tokenData.resource !== "string" || tokenData.resource === "")) {
      throw new BasecampError("api_error", "Token response resource must be a non-empty string when present", {
        httpStatus: response.status,
      });
    }

    return {
      accessToken: tokenData.access_token,
      // `?? undefined`: JSON null is legal on the wire for the optional
      // fields (absent per SPEC) — never leak null through the public type.
      refreshToken: tokenData.refresh_token ?? undefined,
      tokenType: tokenData.token_type ?? "Bearer",
      expiresIn: tokenData.expires_in,
      expiresAt: tokenData.expires_in
        ? new Date(Date.now() + tokenData.expires_in * 1000)
        : undefined,
      scope: tokenData.scope ?? undefined,
      resource: tokenData.resource ?? undefined,
    };
  } catch (err) {
    if (err instanceof BasecampError) {
      throw err;
    }

    if (err instanceof Error) {
      if (err.name === "AbortError") {
        throw new BasecampError("network", "Token request timed out", {
          cause: err,
          retryable: true,
        });
      }

      throw new BasecampError("network", `Token request failed: ${err.message}`, {
        cause: err,
        retryable: true,
      });
    }

    throw new BasecampError("network", "Token request failed with unknown error", {
      retryable: true,
    });
  } finally {
    clearTimeout(timeoutId);
  }
}

/**
 * Checks if a token is expired or about to expire.
 *
 * @param token - The token to check
 * @param bufferSeconds - Buffer time before actual expiration (default: 60)
 * @returns true if the token is expired or will expire within the buffer time
 *
 * @example
 * ```ts
 * if (isTokenExpired(token)) {
 *   token = await refreshToken({ ... });
 * }
 * ```
 */
export function isTokenExpired(token: OAuthToken, bufferSeconds = 60): boolean {
  if (!token.expiresAt) {
    // No expiration info - assume not expired
    return false;
  }

  const bufferMs = bufferSeconds * 1000;
  return Date.now() + bufferMs >= token.expiresAt.getTime();
}

File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/sdk/oauth/Exchange.kt
package com.basecamp.sdk.oauth

import io.ktor.client.*
import io.ktor.client.plugins.*
import io.ktor.client.request.*
import io.ktor.client.request.forms.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import com.basecamp.sdk.BasecampException
import com.basecamp.sdk.requireSecureEndpoint
import com.basecamp.sdk.http.currentTimeMillis

/**
 * OAuth access token response.
 */
data class OAuthToken(
    val accessToken: String,
    val refreshToken: String?,
    val tokenType: String,
    val expiresIn: Long?,
    /** Computed wall-clock expiration time (epoch milliseconds). */
    val expiresAt: Long?,
    val scope: String?,
    /**
     * RFC 8707 resource indicator the token is bound to (BC5:
     * `urn:bc:account:<id>`). Echo it as the `resource` parameter when
     * refreshing — BC5 multi-account refresh tokens reject a refresh without
     * it (SPEC §16). Appended last: earlier parameters keep their positions.
     */
    val resource: String? = null,
)

@Serializable
internal data class RawTokenResponse(
    @SerialName("access_token") val accessToken: String,
    @SerialName("refresh_token") val refreshToken: String? = null,
    @SerialName("token_type") val tokenType: String? = null,
    @SerialName("expires_in") val expiresIn: Long? = null,
    val scope: String? = null,
    val resource: String? = null,
)

@Serializable
internal data class OAuthErrorResponse(
    val error: String,
    @SerialName("error_description") val errorDescription: String? = null,
)

private val tokenJson = Json { ignoreUnknownKeys = true }
private const val MAX_RESPONSE_SIZE = 1_048_576L // 1 MB

/** Bounded per-request timeout for every token-endpoint POST — the 30 s credential-POST default shared across the SDKs (SPEC §16). */
private const val TOKEN_REQUEST_TIMEOUT_MS = 30_000L

/**
 * The redirects the token endpoint refuses outright (SPEC §16 "Token-Endpoint
 * Transport Policy") — the same set the signed download hop refuses (#809).
 * 304 is deliberately absent: it is a cache validator with no `Location`, and
 * falls through to the generic non-success branch below.
 */
private val REDIRECT_STATUSES = setOf(301, 302, 303, 307, 308)

/**
 * Builds a hardened HTTP client for token-endpoint POSTs: redirects suppressed
 * ([HttpClient.followRedirects] = false, so a 3xx is classified below rather
 * than any engine chasing an attacker-influenced `Location` with the
 * credentials re-POSTed) and a bounded per-request timeout ([HttpTimeout]) so
 * a stalled token request cannot hang an exchange or refresh unbounded.
 *
 * When [baseClient] is supplied its engine is reused but wrapped so the
 * hardening applies regardless — redirect suppression is a security
 * invariant, not a default (the device flow does the same). The returned
 * wrapper is always closed by the caller and, because Ktor only closes
 * engines it created, the borrowed engine survives.
 */
private fun hardenedTokenClient(baseClient: HttpClient?): HttpClient {
    val engine = baseClient?.engine
    return if (engine != null) {
        HttpClient(engine) {
            followRedirects = false
            expectSuccess = false
            install(HttpTimeout) { requestTimeoutMillis = TOKEN_REQUEST_TIMEOUT_MS }
        }
    } else {
        HttpClient {
            followRedirects = false
            expectSuccess = false
            install(HttpTimeout) { requestTimeoutMillis = TOKEN_REQUEST_TIMEOUT_MS }
        }
    }
}

/**
 * Exchanges an authorization code for tokens.
 *
 * Supports both standard OAuth 2.0 (`grant_type=authorization_code`) and
 * Basecamp's Launchpad legacy format (`type=web_server`).
 *
 * ```kotlin
 * val token = exchangeCode(
 *     tokenEndpoint = config.tokenEndpoint,
 *     code = authorizationCode,
 *     redirectUri = "https://myapp.com/callback",
 *     clientId = clientId,
 *     clientSecret = clientSecret,
 *     codeVerifier = pkce.verifier,
 * )
 * ```
 */
suspend fun exchangeCode(
    tokenEndpoint: String,
    code: String,
    redirectUri: String,
    clientId: String,
    clientSecret: String,
    codeVerifier: String? = null,
    useLegacyFormat: Boolean = false,
    client: HttpClient? = null,
): OAuthToken {
    val params = if (useLegacyFormat) {
        parametersOf(
            "type" to listOf("web_server"),
            "code" to listOf(code),
            "redirect_uri" to listOf(redirectUri),
            "client_id" to listOf(clientId),
            "client_secret" to listOf(clientSecret),
        )
    } else {
        val map = mutableMapOf(
            "grant_type" to listOf("authorization_code"),
            "code" to listOf(code),
            "redirect_uri" to listOf(redirectUri),
            "client_id" to listOf(clientId),
            "client_secret" to listOf(clientSecret),
        )
        if (codeVerifier != null) {
            map["code_verifier"] = listOf(codeVerifier)
        }
        parametersOf(map)
    }

    return postTokenRequest(tokenEndpoint, params, client)
}

/**
 * Refreshes an access token using a refresh token.
 *
 * ```kotlin
 * val newToken = refreshToken(
 *     tokenEndpoint = config.tokenEndpoint,
 *     refreshToken = currentToken.refreshToken!!,
 *     clientId = clientId,
 *     clientSecret = clientSecret,
 * )
 * ```
 */
suspend fun refreshToken(
    tokenEndpoint: String,
    refreshToken: String,
    clientId: String,
    // Nullable: the public `basecamp-cli` client is a public OAuth client
    // (`token_endpoint_auth_method: none`) and sends no secret.
    clientSecret: String? = null,
    useLegacyFormat: Boolean = false,
    client: HttpClient? = null,
    // RFC 8707 resource indicator, sent only when set. Echo the stored
    // token's resource: BC5 multi-account refresh tokens hard-require it
    // (SPEC §16). Appended last: earlier parameters keep their positions.
    resource: String? = null,
): OAuthToken {
    val params = Parameters.build {
        if (useLegacyFormat) {
            append("type", "refresh")
        } else {
            append("grant_type", "refresh_token")
        }
        append("refresh_token", refreshToken)
        append("client_id", clientId)
        if (!clientSecret.isNullOrEmpty()) append("client_secret", clientSecret)
        if (!resource.isNullOrEmpty()) append("resource", resource)
    }

    return postTokenRequest(tokenEndpoint, params, client)
}

/**
 * Checks whether a token is expired (or within the buffer window).
 *
 * @param bufferSeconds Seconds before actual expiration to consider expired (default: 60).
 */
fun isTokenExpired(token: OAuthToken, bufferSeconds: Long = 60): Boolean {
    val expiresAt = token.expiresAt ?: return false // No expiration info → assume valid
    return currentTimeMillis() >= (expiresAt - bufferSeconds * 1000)
}

private suspend fun postTokenRequest(
    endpoint: String,
    params: Parameters,
    client: HttpClient?,
): OAuthToken {
    // Never POST credentials over cleartext (localhost exempt for dev/test).
    requireSecureEndpoint(endpoint, "token endpoint")

    val httpClient = hardenedTokenClient(client)

    try {
        val response = httpClient.submitForm(endpoint, params) {
            accept(ContentType.Application.Json)
        }

        val status = response.status.value

        // Status-first: a refused redirect is classified BEFORE any body read,
        // so a 3xx that drip-feeds its body cannot degrade into a timeout. The
        // hardened client never follows, and the `Location` is never dialled —
        // the endpoint the caller (or discovered metadata) named is the one
        // destination these credentials go to (SPEC §16).
        if (status in REDIRECT_STATUSES) {
            throw BasecampException.Api(
                "redirect $status on the token endpoint is not followed",
                httpStatus = status,
            )
        }

        val body = response.bodyAsText()

        if (body.length > MAX_RESPONSE_SIZE) {
            throw BasecampException.Api(
                "OAuth token response exceeds size limit",
                httpStatus = status,
            )
        }

        if (!response.status.isSuccess()) {
            val errorResp = runCatching { tokenJson.decodeFromString<OAuthErrorResponse>(body) }.getOrNull()
            val message = errorResp?.errorDescription
                ?: errorResp?.error
                ?: "Token request failed: HTTP $status"
            throw BasecampException.Auth(
                message = BasecampException.truncateMessage(message),
            )
        }

        // A token response that fails to decode may still contain credential
        // material, and kotlinx embeds input excerpts in its exception
        // messages — map to a status-only fault (no cause: cause messages
        // surface in stack traces) instead of propagating it.
        val raw = runCatching { tokenJson.decodeFromString<RawTokenResponse>(body) }.getOrElse {
            throw BasecampException.Api("Failed to parse token response", httpStatus = status)
        }
        // resource: absent and JSON null decode to null (unset); when present
        // it must be non-empty (SPEC §16) — an empty binding is not a binding.
        // A non-string resource fails deserialization above.
        if (raw.resource != null && raw.resource.isEmpty()) {
            throw BasecampException.Api(
                "Token response resource must be a non-empty string when present",
                httpStatus = status,
            )
        }
        // A 2xx with an EMPTY access_token is malformed, not a success —
        // matching the device flow and the other SDKs' non-empty contract.
        if (raw.accessToken.isEmpty()) {
            throw BasecampException.Api(
                "Token response missing access_token",
                httpStatus = status,
            )
        }

        val now = currentTimeMillis()
        val expiresAt = raw.expiresIn?.let { now + it * 1000 }

        // token_type: absent/JSON-null defaults to Bearer; a present-but-empty
        // value is malformed (SPEC §16) — matching the device-flow parser.
        val tokenType = raw.tokenType?.also {
            if (it.isEmpty()) {
                throw BasecampException.Api(
                    "Token response token_type must be a non-empty string when present",
                    httpStatus = status,
                )
            }
        } ?: "Bearer"

        return OAuthToken(
            accessToken = raw.accessToken,
            refreshToken = raw.refreshToken,
            tokenType = tokenType,
            expiresIn = raw.expiresIn,
            expiresAt = expiresAt,
            scope = raw.scope,
            resource = raw.resource,
        )
    } catch (e: HttpRequestTimeoutException) {
        // The wrapper's HttpTimeout fired. Mapped explicitly — it subclasses
        // CancellationException, so left alone it would masquerade as a
        // cooperative cancellation — to the retryable network fault the other
        // SDKs raise here ("Token request timed out", TS/Python/Ruby).
        throw BasecampException.Network("Token request timed out", cause = e)
    } finally {
        // Always ours: hardenedTokenClient built it, an injected client only
        // lent its engine (which close() leaves running).
        httpClient.close()
    }
}
